PR: https://github.com/odoo/odoo/pull/253079

From: 3259a7fae4ca4ee98f7e02da5109c43cd9e77f69
From: Tatiana Skvortsova (taskv)
Date: 2026-04-13 06:36:16

Breaking data model changes scores: del:18 + add:0, change matches:
-    departure_action_at_departure = fields.Boolean(related='version_id.departure_action_at_departure', inherited=True,
-    departure_action_other_date = fields.Date(related='version_id.departure_action_other_date', inherited=True,
-    departure_do_archive_employee = fields.Boolean(related='version_id.departure_do_archive_employee', inherited=True,
-    departure_do_archive_user = fields.Boolean(related='version_id.departure_do_archive_user', inherited=True,
-    departure_do_set_date_end = fields.Boolean(related='version_id.departure_do_set_date_end', inherited=True,
-    departure_has_selected_actions = fields.Boolean(related='version_id.departure_has_selected_actions', inherited=True,
-    action_at_departure = fields.Boolean(string="Action at", default=True)
-    action_other_date = fields.Date(string="Apply date")
-    do_archive_user = fields.Boolean(
-    do_archive_employee = fields.Boolean(default=True, string="Archive Employee")
-    do_set_date_end = fields.Boolean(
-    has_selected_actions = fields.Boolean(compute='_compute_has_selected_actions')
-    departure_action_at_departure = fields.Boolean(related='departure_id.action_at_departure', readonly=False, groups="hr.group_hr_user")
-    departure_action_other_date = fields.Date(related='departure_id.action_other_date', readonly=False, groups="hr.group_hr_user")
-    departure_do_archive_employee = fields.Boolean(related='departure_id.do_archive_employee', readonly=False, groups="hr.group_hr_user")
-    departure_do_archive_user = fields.Boolean(related='departure_id.do_archive_user', readonly=False, groups="hr.group_hr_user")
-    departure_do_set_date_end = fields.Boolean(related='departure_id.do_set_date_end', readonly=False, groups="hr.group_hr_user")
-    departure_has_selected_actions = fields.Boolean(related='departure_id.has_selected_actions', groups="hr.group_hr_user")

Total Changes: 112

[IMP] l10n_be_hr_payroll: Improve End of Collaboration UX

Improve the UX of the "End of Collaboration" procedure, particularly for Belgium.

The option to select which actions are executed upon employee departure
(delete user, archive employee, unassign car, etc.) has been removed.

All actions are now executed automatically. However, it is still possible
to define the date on which they will be applied.

Related PRs:
Enterprise: odoo/enterprise#110147
Upgrade: odoo/upgrade#9653

task-5958676

test

closes odoo/odoo#253079

Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/hr_employee.py
+++ b/addons/hr/models/hr_employee.py
@@ -288,18 +288,8 @@ class HrEmployee(models.Model):
         readonly=False, groups="hr.group_hr_user")
     departure_date = fields.Date(related='version_id.departure_date', inherited=True,
         readonly=False, groups="hr.group_hr_user")
-    departure_action_at_departure = fields.Boolean(related='version_id.departure_action_at_departure', inherited=True,
+    departure_action_date = fields.Date(related='version_id.departure_action_date', inherited=True,
         readonly=False, groups="hr.group_hr_user")
-    departure_action_other_date = fields.Date(related='version_id.departure_action_other_date', inherited=True,
-        readonly=False, groups="hr.group_hr_user")
-    departure_do_archive_employee = fields.Boolean(related='version_id.departure_do_archive_employee', inherited=True,
-        readonly=False, groups="hr.group_hr_user")
-    departure_do_archive_user = fields.Boolean(related='version_id.departure_do_archive_user', inherited=True,
-        readonly=False, groups="hr.group_hr_user")
-    departure_do_set_date_end = fields.Boolean(related='version_id.departure_do_set_date_end', inherited=True,
-        readonly=False, groups="hr.group_hr_user")
-    departure_has_selected_actions = fields.Boolean(related='version_id.departure_has_selected_actions', inherited=True,
-        groups="hr.group_hr_user")
     departure_apply_immediately = fields.Boolean(related='version_id.departure_apply_immediately', inherited=True,
         groups="hr.group_hr_user")
     departure_apply_date = fields.Date(related='version_id.departure_apply_date', inherited=True,
@@ -2181,15 +2171,22 @@ class HrEmployee(models.Model):
 
     def action_new_departure(self):
         self.ensure_one()
+        if not self.is_in_contract:
+            raise UserError(self.env._("You can't end the collaboration of an employee without contract, archive it instead."))
         if self.departure_id:
             if self.departure_id.apply_date:
                 raise UserError(self.env._("You can't modify the departure of an employee that has already departed."))
             return {
-                'name': self.env._('End of collaboration'),
-                'res_model': 'hr.employee.departure',
-                'res_id': self.departure_id.id,
-                'type': 'ir.actions.act_window',
-                'view_mode': 'form',
+                "type": "ir.actions.client",
+                "tag": "departure_conflict_dialog",
+                "params": {
+                    "title": self.env._("Departure already configured"),
+                    "message": self.env._(
+                        "Employee is already configured in a Departure process. \n"
+                        "Cancel it before configurating a new End of Collaboration."
+                    ),
+                    "employee_id": self.id,
+                },
             }
         return {
             'name': self.env._('End of collaboration'),

--- a/addons/hr/models/hr_employee_departure.py
+++ b/addons/hr/models/hr_employee_departure.py
@@ -1,4 +1,5 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
+from dateutil.relativedelta import relativedelta
 
 from odoo import api, fields, models
 from odoo.exceptions import ValidationError
@@ -34,57 +35,46 @@ class HrEmployeeDeparture(models.Model):
         help="Date at which the departure process starts. Differs from the actual departure date in case of a notice period.")
     departure_date = fields.Date(string="Departure Date", compute="_compute_departure_date",
         store=True, readonly=False, help="Date at which the departure actually takes place.")
-    action_at_departure = fields.Boolean(string="Action at", default=True)
-    action_other_date = fields.Date(string="Apply date")
+    action_date = fields.Date(string="Archive Employee On", compute="_compute_action_date",
+        store=True, help="Date at which the departure actually takes place.")
     is_user_employee = fields.Boolean(
         compute='_compute_is_user_employee',
         export_string_translation=False,
     )
-    do_archive_user = fields.Boolean(
-        string="Archive Related User",
-        help="""
-            If checked, the related user will be removed from the system.
-            The user will however not be removed if they still have an employee in any other company after this departure.
-        """,
-    )
-    do_archive_employee = fields.Boolean(default=True, string="Archive Employee")
-    do_set_date_end = fields.Boolean(
-        string="Set Contract End Date", default=True,
-        help="Set the departure date as the contract end date and delete all future versions.")
-    has_selected_actions = fields.Boolean(compute='_compute_has_selected_actions')
     apply_immediately = fields.Boolean(compute="_compute_apply_immediately")
     apply_date = fields.Date(readonly=True)
 
-    def _get_action_fields(self):
-        return [f for f in self._fields if f.startswith('do_')]
-
     @api.depends('dismissal_date')
     def _compute_departure_date(self):
         # meant to be overriden in case of notice period
         for departure in self:
             departure.departure_date = departure.dismissal_date
 
+    @api.depends('departure_date')
+    def _compute_action_date(self):
+        for departure in self:
+            if departure.action_date and departure.action_date < departure.departure_date:
+                departure.action_date = departure.departure_date + relativedelta(days=1)
+
+    @api.onchange("departure_date")
+    def _onchange_departure_date(self):
+        self.action_date = self.departure_date + relativedelta(days=1)
+
     @api.depends('employee_id.user_id')
     def _compute_is_user_employee(self):
         for departure in self:
             departure.is_user_employee = bool(departure.employee_id.user_id)
 
-    @api.depends(lambda self: self._get_action_fields())
-    def _compute_has_selected_actions(self):
-        action_fields = self._get_action_fields()
-        for departure in self:
-            departure.has_selected_actions = any(departure[field] for field in action_fields)
-
-    @api.depends('action_at_departure', 'departure_date', 'action_other_date')
+    @api.depends('departure_date', 'action_date')
     def _compute_apply_immediately(self):
         today = fields.Date.today()
         for departure in self:
             if not departure.departure_date:
                 departure.apply_immediately = False
-            elif departure.action_at_departure:
-                departure.apply_immediately = departure.departure_date <= today
+            elif departure.action_date:
+                departure.apply_immediately = departure.action_date <= today
             else:
-                departure.apply_immediately = not departure.action_other_date or departure.action_other_date <= today
+                departure.apply_immediately = departure.departure_date < today
 
     @api.constrains('employee_id')
     def _check_departure_validity(self):
@@ -124,14 +114,10 @@ class HrEmployeeDeparture(models.Model):
         departures = self.search([
             ('apply_date', '=', False),
             '|',
+                ('action_date', '<=', today),
                 '&',
-                    ('action_at_departure', '=', True),
-                    ('departure_date', '<=', today),
-                '&',
-                    ('action_at_departure', '=', False),
-                    '|',
-                        ('action_other_date', '<=', today),
-                        ('action_other_date', '=', False),
+                    ('action_date', '=', False),
+                    ('departure_date', '<', today),
         ])
         departures.action_register()
 
@@ -158,12 +144,8 @@ class HrEmployeeDeparture(models.Model):
             if not user:
                 continue
             active_user_employees = user.employee_ids.filtered('active')
-            if not user.active or\
-                    (active_user_employees - departures.employee_id) or\
-                    any(not d.do_archive_user for d in departures):
-                # We don't archive the user:
-                # - if all related active employees are not departing
-                # - if any of the departures related asks to keep it active
+            if not user.active or (active_user_employees - departures.employee_id):
+                # We don't archive the user if all related active employees are not departing
                 users_to_keep_active += user
                 continue
             users_to_archive += user
@@ -173,22 +155,18 @@ class HrEmployeeDeparture(models.Model):
         emp_to_archive = self.env['hr.employee']
         for departure in self:
             employee = departure.employee_id
-            if departure.action_at_departure:
-                apply_date = departure.departure_date
-            else:
-                apply_date = departure.action_other_date or fields.Date.today()
+            apply_date = departure.action_date or departure.departure_date + relativedelta(days=1) or fields.Date.today()
             if apply_date > fields.Date.today():
                 raise ValidationError(self.env._(
                     "The apply date isn't reached yet for the departure of %(emp)s.",
                     emp=departure.employee_id.name))
 
-            if departure.do_archive_employee and departure.employee_id.active:
+            if departure.employee_id.active:
                 emp_to_archive += employee
 
-            if departure.do_set_date_end:
-                if employee.sudo().contract_date_start:
-                    employee.sudo().write({'contract_date_end': departure.departure_date})
-                employee.version_ids.filtered(lambda v: v.date_version > departure.departure_date).unlink()
+            if employee.sudo().contract_date_start:
+                employee.sudo().write({'contract_date_end': departure.departure_date})
+            employee.version_ids.filtered(lambda v: v.date_version > departure.departure_date).unlink()
 
         emp_to_archive.action_archive()
         self.apply_date = fields.Date.today()

--- a/addons/hr/models/hr_version.py
+++ b/addons/hr/models/hr_version.py
@@ -129,12 +129,7 @@ class HrVersion(models.Model):
     departure_description = fields.Html(related='departure_id.departure_description', readonly=False, groups="hr.group_hr_user")
     dismissal_date = fields.Date(related='departure_id.dismissal_date', readonly=False, groups="hr.group_hr_user", tracking=1)
     departure_date = fields.Date(related='departure_id.departure_date', readonly=False, groups="hr.group_hr_user", tracking=1)
-    departure_action_at_departure = fields.Boolean(related='departure_id.action_at_departure', readonly=False, groups="hr.group_hr_user")
-    departure_action_other_date = fields.Date(related='departure_id.action_other_date', readonly=False, groups="hr.group_hr_user")
-    departure_do_archive_employee = fields.Boolean(related='departure_id.do_archive_employee', readonly=False, groups="hr.group_hr_user")
-    departure_do_archive_user = fields.Boolean(related='departure_id.do_archive_user', readonly=False, groups="hr.group_hr_user")
-    departure_do_set_date_end = fields.Boolean(related='departure_id.do_set_date_end', readonly=False, groups="hr.group_hr_user")
-    departure_has_selected_actions = fields.Boolean(related='departure_id.has_selected_actions', groups="hr.group_hr_user")
+    departure_action_date = fields.Date(related='departure_id.action_date', readonly=False, groups="hr.group_hr_user")
     departure_apply_immediately = fields.Boolean(related='departure_id.apply_immediately', groups="hr.group_hr_user")
     departure_apply_date = fields.Date(related='departure_id.apply_date', groups="hr.group_hr_user")
 
